### ISSUE TITLE
Add custom metadata to annotators

### ISSUE THREAD

**Issue by renogomez** [JohnSnowLabs/spark-nlp#920]
**TL;DR**
If there is a way to modify a column and keep it as a valid annotator then the feature is not necessary (my apologies and please provide and example on how to do it, that would be powerful). Refer to the last part (alternatives)

**Is your feature request related to a problem? Please describe.**

I have been using TextMatcher and Regex annotator to extract information from sentences. RegexMatcher metadata has an identifier which allows me to know exactly from which pattern I got one each specific annotation (identifier field). However, something similar is not possible with TextMatcher annotator. It becomes a problem if you want to merge the annotations to process all the ones belonging to one specific sentence for instance. Consider the output columns from 3 different annotators

```
color chunk (TextMatcher)
|[[chunk, 4, 6, Red, [sentence -> 0, chunk -> 0], []]]
color adj chunk (TextMatcher)
|[[chunk, 8, 12, Light, [sentence -> 0, chunk -> 0], []]]
interval chunk (RegextMatcher)
|[[chunk, 60, 76, -2257Km to 2,29Km, [identifier -> INTERVAL, sentence -> 0, chunk -> 0], []], [chunk, 61, 76, 2257Km to 2,29Km, [identifier -> INTERVAL, sentence -> 0, chunk -> 1], []]]
```

If I later use a chunk2Doc tranformer like this:

```
chunk_doc = Chunk2Doc()\
    .setInputCols(["color_span", "color_adj_span", "interval_span"])\
    .setOutputCol("doc_spans")\
```

In the result I won't be able to know from which dictionary came one annotation: 

```
[
[chunk, 4, 6, Red, [sentence -> 0, chunk -> 0], []],
[chunk, 8, 12, Light, [sentence -> 0, chunk -> 0], []]
[chunk, 60, 76, -2257Km to 2,29Km, [identifier -> INTERVAL, sentence -> 0, chunk -> 0], []], 
[chunk, 61, 76, 2257Km to 2,29Km, [identifier -> INTERVAL, sentence -> 0, chunk -> 1], []]
]
```

**Describe the solution you'd like**
I would like to have the possibility to specify a metadata field to the annotators or at least something similar to the RegexMatcher. In other words: _add a param for custom metadata to be added as a key/value to be set and added to the annotator metadata which can be used in this case as source->COLOR_DICT for instance._

```
[chunk, 4, 6, Red, [source -> COLOR_DICT, sentence -> 0, chunk -> 0], []],
[chunk, 8, 12, Light, [source -> ADJ_DICT, sentence -> 0, chunk -> 0], []]
```

It would be great if this is consistent in all annotator.

**Describe alternatives you've considered**
Instead of using the Chunk2Doc to the chunk annotator I got previously, I considered unified them using a UDF

```
def get_annotator_list(arr):
    labels_from_dict = list()
    for ent in arr[0]:
        ent_token = Annotation(
                ent.annotatorType,
                ent.begin,
                ent.end,
                ent.result,
                {'identitifier': 'dict_entities', 
                 'sentence': ent.metadata['sentence'],
                'chunk': ent.metadata['chunk']},
                ent.embeddings)
        labels_from_dict.append(ent_token)
    return labels_from_dict
to_annotation_list = F.udf(get_annotator_list, Annotation.arrayType())

```
But the problem is that the output its not recognize as a valid annotator anymore. I think this is true for any modification outside the library, I guess for consistency reasons. If there is another way to modify a column and keep it as a valid annotator then the feature is not necessary (and please provide and example on how to do it, that would be powerful)

**Additional context**
None

**Comment by maziyarpanahi** [comment]
This feature was added in 2.5.2 release. Thanks for the feature request.
